docs/programming/PerformanceCounters.md

```diff
@@ -0,0 +1,695 @@
+
+```
+From the output we can see:
+
+- Block Size (256) × Grid Size (4) = 1024 total threads. This is the total parallel workload for the  kernel.
+- 108 SMs: A100 GPU has 108 compute units available.
+- Waves Per SM = 0.00: Each SM received less than one full wave of work, which is not enough to keep it busy.
+- Compute Throughput = 0.04%: The GPU was virtually idle during the kernel execution.
+
+These numbers clearly show that the kernel is too small to utilize the hardware effectively.
+
+| Symptom               | Likely Cause                          | Recommended Fix                           | 
+| :----------------- | :------------------------------------------- |:-------------------------------------------------- |
+| Grid size is smaller than #SMs	| Not enough blocks to occupy all compute units	| Increase the number of blocks by enlarging the problem size |
+| SM throughput is very low (< 10%)	| GPU is mostly idle	| Optimize workload or parallelism granularity |
+| Waves Per SM is 0	| Too little parallel work	| Increase N or restructure the launch configuration |
+
+In the original kernel, the launch configuration was:
+```
+#define N 1024
+dim3 block(256);
+dim3 grid((N + block.x - 1) / block.x);  // → grid.x = 4 blocks
+```
+This launches just 4 blocks, causing underutilization.
+To improve utilization, update the launch configuration to the following:
+```
+#define N (1 << 20)  // 1,048,576 elements
+dim3 block(256);
+dim3 grid((N + block.x - 1) / block.x);  // → grid.x = 4096 blocks
+```
+This updated configuration allows the GPU to schedule multiple waves of work across all SMs, leading to much better throughput and performance.
+
+## Nsight Systems (nsys)
+
+Nsight Systems  is a system-wide profiler that traces the interactions between CPU and GPU, memory transfers, and OS-level activity. It’s useful for understanding:
+
+- Launch latency of GPU kernels.
+
+- Investigate host-device memory transfers.
+
+- Identify CPU-side inefficiencies (e.g., blocking, thread stalls)
+
+### Getting Started
+
+To use ```nsys```, load the appropriate CUDA module and invoke ```nsys``` by prefixing it to your CUA application:
+
+```
+$ module load cuda
+$ nsys profile --trace=cuda,osrt ./vectorAdd
+```
+- ```--trace=cuda```: Capture CUDA kernel launches and memory transfers.
+
+- ```--trace=osrt```: Trace OS runtime activity (threads, processes, etc.).
+
+This generates a report file in binary format:
+```
+./report.nsys-rep
+```
+To view a quick CLI summary of the captured data:
+```
+$ nsys stats report.nsys-rep
+```
+
+::::{dropdown} Click here to view full output 
+:icon: note
+
+Generating SQLite file report.sqlite from report.nsys-rep
```

Comment:
I think you should make the output in this drop down a code block. Otherwise, the formatting looks super strange.